[wip] Cli args in scripts

tpasternak opened the pull request.

tpasternak commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -211,7 +230,9 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
     )
     inputs.fromRoot { root =>
-      val output = os.proc(TestUtil.cli, extraOptions, "dir", "--main-class", "print").call(cwd =
+      // TODO: It's probably impossible to handle `--main-class print`, because we put main method in the other
```
@alexarchambault @lwronski I couldn't find any way to fix it. This means the scripts users will have to pass `--main-class <somethnig>_sc`, or we will have to add `_sc` suffix silently in options engine

alexarchambault replied:
`--main-class print_sc` is fine for me. I wish it could have been `print.sc`, with the right back-ticks added for it to work.

tpasternak replied:
I can do it in a separate PR. The problem is that we don't support special characters in --main-class at all. `scala.sc` class is translated to `scala$u002Esc`, but arg passed to `--main-class` remains unchanged. We would have find the code in scala compiler that is used to transform the name and apply this also here. As a quick hack we can just translate some arbitrary symbols (dot, space etc.)

alexarchambault commented on modules/build/src/main/scala/scala/build/Build.scala:
```diff
@@ -203,6 +203,16 @@ object Build {
+      case Some(v) if v.startsWith("3")    => false
+      case Some(v) if v.startsWith("2.13") => true
+      case Some(v) if v.startsWith("2.12") =>
+        !inputs.sourceFiles().exists(_.isInstanceOf[Inputs.ScriptBase])
```
Let's not use `isInstanceOf` (if ever the type returned by `sourceFile` changes, we'll get no compiler error here). Better use a match:
```scala
        !inputs.sourceFiles().exists {
          case _: Inputs.ScriptBase => true
          case _ => false
        }
```

alexarchambault commented on modules/build/src/main/scala/scala/build/Inputs.scala:
```diff
@@ -127,6 +127,8 @@ object Inputs {
 
   sealed trait SingleElement extends Element
 
+  sealed trait ScriptBase
```
`AnyScript` instead of `ScriptBase`? (like the existing `AnyScalaFile`)

alexarchambault commented on modules/build/src/main/scala/scala/build/errors/ScalaNativeCompatibilityError.scala:
```diff
@@ -0,0 +1,10 @@
+package scala.build.errors
+
+case object ScalaNativeCompatibilityError
```
Exceptions shouldn't be case classes or case objects. For case objects, this gives them a wrong stack trace. For case classes, this breaks the case class semantics, see the discussion around https://github.com/scalameta/metals/pull/3145#discussion_r712800065.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -713,6 +713,32 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+    )
+    inputs.fromRoot { root =>
+      val p =
+        os.proc(TestUtil.cli, "main0.sc", "f.sc", "--", "20").call(cwd = root, stdout = os.Pipe)
```
`stdout = os.Pipe` is the default, no need to specify it.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -713,6 +713,32 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+      )
+    )
+    inputs.fromRoot { root =>
+      val p = os.proc(TestUtil.cli, "f.sc", "--", "16").call(cwd = root, stdout = os.Pipe)
```
Same as above.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -118,10 +118,29 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
+          cwd =
+            root,
+          check = false,
+          stderr = os.Pipe
```
Same as above.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -118,10 +118,29 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
+      inputs.fromRoot { root =>
+        val output = os.proc(TestUtil.cli, extraOptions, "--native", "a.sc").call(
+          cwd =
+            root,
```
Why a new line?

alexarchambault reviewed (commented):
I didn't look closely at `MainClassRunner` yet… I'll get back to it a bit later.

romanowski commented on modules/build/src/main/scala/scala/build/errors/ScalaNativeCompatibilityError.scala:
```diff
@@ -0,0 +1,10 @@
+    extends BuildException(
+      """scala-cli: invalid option: '--native' for scripts is supported only for scala 2.13.*
+        |Please try one of the following combinations:
+        |  scala-cli --native -S 2.12 <...> (for *.sc files)
```
```suggestion
        |  scala-cli --native -S 2.13 <...> (for *.sc files)
```

tpasternak replied:
For 2.12 we support only .scala files
For 2.13 both .scala and .sc

Hopefully will be fixed in scala-native 4.1

romanowski reviewed (changes requested):
Please include documentation or even better a cookbook

lwronski commented on website/docs/cookbooks/scripting.md:
```diff
@@ -0,0 +1,28 @@
+# Develop Scripts in Scala
```
Maybe, we should merge scripting cookbook with this [one](https://github.com/VirtusLab/scala-cli/blob/master/website/docs/cookbooks/scala-scripts.md).

tpasternak replied:
Can I do it in a follow-up PR?

alexarchambault commented on modules/build/src/main/scala/scala/build/Inputs.scala:
```diff
@@ -127,6 +127,8 @@ object Inputs {
 
   sealed trait SingleElement extends Element
 
+  sealed trait AnyScript
```
NIT Make that extend `Element`, so that we know it's related to it.
```suggestion
  sealed trait AnyScript extends Element
```

alexarchambault commented on modules/build/src/main/scala/scala/build/Build.scala:
```diff
@@ -316,40 +310,54 @@ object Build {
-        bloopServer = bloopServer,
-        crossBuilds = crossBuilds
+  ): Either[BuildException, Builds] =
+    if (options.platform == Platform.Native && !scalaNativeSupported(options, inputs))
```
This check should be done in `buildOnce` rather than here. Some new options may come up from the sources, and can set or change the platform later.

alexarchambault commented on modules/build/src/main/scala/scala/build/Build.scala:
```diff
@@ -316,40 +310,54 @@ object Build {
     classesRootDir(root, projectName) / scope.name
 
+  def scalaNativeSupported(options: BuildOptions, inputs: Inputs): Boolean =
+    options.scalaOptions.scalaVersion match {
```
Getting the Scala version via `options.scalaParams` would give just a `String`, rather than an `Option[String]`, and should be the way to get the Scala version from `BuildOptions`. We may change the default Scala version to a `2.13` one, and `options.scalaParams` would reflect that.

`options.scalaParams` returns an either, but you can use `either` / `value` to deal with it, like other methods in this file.

alexarchambault commented on modules/build/src/main/scala/scala/build/errors/ScalaNativeCompatibilityError.scala:
```diff
@@ -0,0 +1,10 @@
+
+final class ScalaNativeCompatibilityError
+    extends BuildException(
+      """scala-cli: invalid option: '--native' for scripts is supported only for scala 2.13.*
```
**For later**: we shouldn't mention command-line options from `BuildException` message. Options are managed from the `cli` module, that the `build` module knows nothing about. And these options might not be available for all commands.

I'd like to add a way in each command to print a specific message if some particular exceptions happen, to suggest options there.

Also, the "name" of the application should be `(new scala.cli.internal.Argv0).get("scala-cli")` rather than just `scala-cli`. The former relies on more idiomatic native stuff. If the `scala-cli` file is renamed to `scala` or `scala-cli-1.0`, its value will be `scala` or `scala-cli-1.0` for example.

We should be be able to handle that if we print specific messages for some exceptions, like I mentioned just above.

alexarchambault commented on modules/build/src/main/scala/scala/build/internal/CustomClassCodeWrapper.scala:
```diff
@@ -1,135 +0,0 @@
-package scala.build.internal
```
NIT Could this class be removed in a specific commit, doing only that?

alexarchambault commented on modules/build/src/test/scala/scala/build/tests/TestUtil.scala:
```diff
@@ -17,7 +17,7 @@ object TestUtil {
       val generated0 = generated()
       assert(
-        generated0.map(_.toString).toSet == expected.toSet, {
+        expected.toSet == generated0.map(_.toString).toSet, {
```
This doesn't change anything, does it?

alexarchambault commented on modules/cli/src/main/scala/scala/cli/commands/SharedOptions.scala:
```diff
@@ -107,10 +107,6 @@ final case class SharedOptions(
   def logger = logging.logger
 
-  private def codeWrapper: Option[CodeWrapper] =
-    if (classWrap) Some(CustomCodeClassWrapper)
```
I think we should remove the `classWrap` option here, and the associated CLI option too if there's one.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -713,6 +713,32 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
     }
   }
+
+  test("interconnection between scripts") {
```
Why is this in `BspTestDefinitions`? `RunTestDefinitions` would be a better fit for this.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:
```diff
@@ -25,7 +25,7 @@ abstract class PackageTestDefinitions(val scalaVersionOpt: Option[String])
     val launcherName = {
       val ext = if (Properties.isWin) ".bat" else ""
-      fileName.stripSuffix(".sc") + ext
+      fileName.stripSuffix(".sc") + "_sc" + ext
```
We shouldn't add a `_sc` to the default package name. Maybe it could just be stripped in the `package` command itself?

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:
```diff
@@ -85,7 +86,7 @@ abstract class PackageTestDefinitions(val scalaVersionOpt: Option[String])
       )
     )
-    val destName = fileName.stripSuffix(".sc") + ".js"
+    val destName = fileName.stripSuffix(".sc") + "_sc.js"
```
Same as above, no `_sc` in the default output file name.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:
```diff
@@ -125,7 +126,7 @@ abstract class PackageTestDefinitions(val scalaVersionOpt: Option[String])
     val destName = {
       val ext = if (Properties.isWin) ".exe" else ""
-      fileName.stripSuffix(".sc") + ext
+      fileName.stripSuffix(".sc") + "_sc" + ext
```
Same as above, no `_sc`.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala:
```diff
@@ -162,7 +163,7 @@ abstract class PackageTestDefinitions(val scalaVersionOpt: Option[String])
       )
     )
-    val launcherName = fileName.stripSuffix(".sc") + ".jar"
+    val launcherName = fileName.stripSuffix(".sc") + "_sc.jar"
```
Same as above, no `_sc`.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -118,10 +118,26 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
+        )
+      )
+      inputs.fromRoot { root =>
+        val output = os.proc(TestUtil.cli, extraOptions, "--native", "a.sc").call(
```
Maybe we should just default to `2.13` for Scala Native, and not fail that particular command…

alexarchambault commented:
I pushed an extra commit getting rid of the `MainClassRunner` and `_trait` stuff, that I think are unnecessary (and make scala-cli generate too many classes for single basic scripts.)

tpasternak commented:
Will it work for files named `main.sc`? Btw scala compiler escapes dot symbols with $002e, we would have to escape them also in main class detection code

alexarchambault commented:
> Will it work for files named `main.sc`?

Good point, I think adding the package name when we refer to that object from the `main` method should do.

alexarchambault commented:
> Btw scala compiler escapes dot symbols with $002e, we would have to escape them also in main class detection code

We can try that later, it might not be trivial overall.

alexarchambault commented:
> Good point, I think adding the package name when we refer to that object from the main method should do.

Quickly tried my suggestion ^, and it seems it doesn't work when we're in the root package… (but it works fine else)

alexarchambault commented on modules/build/src/main/scala/scala/build/internal/CustomCodeWrapper.scala:
```diff
@@ -1,23 +1,49 @@
+                             |    argsOpt0 = Some(args)
+                             |  }
+                             |  def argsOpt: Option[Seq[String]] = argsOpt0
+                             |  def args: Seq[String] = argsOpt.getOrElse {
```
Let's keep `Array[String]` instead of `Seq[String]` here. We're getting a warning about the `Array` to `Seq` conversion in Scala 3 else.

alexarchambault replied:
About the issue if the script is called `main.sc`: IMO we should just refuse to have a script named `main.sc` at the root like this. It seems we can't refer to other objects at the root with `_root_.ObjectName` (I ran into this limitation too in the `repl` command…). We could have put the scripts in a default package, but this raises a number of other issues (if we have both `.sc` and `.scala` files at the root, they can't reference each other then…)

We should also probably add some `$` signs to the other variables and methods in this object, so that they don't clash with the script name either (like `args.sc` or `argsOpt.sc`).

tpasternak replied:
Ok, but we already have `main.sc` cases in tests and in documentation

We load this gist in one of the tests https://gist.github.com/alexarchambault/7b4ec20c4033690dd750ffd601e540ec

alexarchambault replied:
We should change that then (I just fixed the gist).

romanowski commented:
It seems that @tpasternak solurion works for now and the only downside is a bit complicated but I am fine with having that in our codebase. 

As for the @alexarchambault concerns: I think we can live with so many generated classes for now.

Let's merge it for now and we can eventually fix it later.

alexarchambault commented:
Why did you revert my commit? The tests only needed minor changes to pass after it.

alexarchambault commented:
The multiple classes isn't the only problem of the `MainClassRunner` stuff. It basically creates two instances of the user variables (one in `new MainClassRunner`, one in the object extending the trait where the user code is put), which runs the entry point script twice in particular:
```text
$ cat test-proj/foo.sc
var messages = List.empty[String]
bar.init()
println("Messages:")
for (m <- messages)
  println(m)

$ cat test-proj/bar.sc
def init(): Unit =
  foo.messages = "Hello" :: foo.messages

$ ./mill -i cli.run test-proj -M foo_sc
Compiling project (Scala 3.0.2, JVM)
Compiled project (Scala 3.0.2, JVM)
Messages:
Hello
Messages:
```

tpasternak commented:
Ok, I reverted it because it did not support main.sc, but you're right my approach was wrong, code cannot be executed twice

tpasternak committed:
Remove CustomCodeClassWrapper

tpasternak committed:
Read CLI args in scripts execution

In order ot achieve this, it was reqired to rediesign the generated
that is used to wrap around scripts. The scipt's body now lands in
<filename>_sc class instead of just <filename>.

This enables referencing symbols within <filename> class from the
other script, as well as execution of the script body

Co-authored-by: Alexandre Archambault <alexandre.archambault@gmail.com>

tpasternak committed:
Change wrapper symbol names to reserved

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -488,34 +506,36 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
       }
-      expect(exceptionLines == expectedLines)
+      expect(exceptionLines.length == expectedLines.length)
+      for { i <- Range(0, exceptionLines.length) } expect(exceptionLines(i) == expectedLines(i))
```
NIT, more idiomatic way of doing this:
```suggestion
      for (i <- 0 until exceptionLines.length) expect(exceptionLines(i) == expectedLines(i))
```

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -553,17 +573,18 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
-      expect(exceptionLines == expectedLines)
+           |$tab... 2 more""".stripMargin.linesIterator.toVector
+      expect(exceptionLines.length == expectedLines.length)
+      for { i <- Range(0, exceptionLines.length) } expect(exceptionLines(i) == expectedLines(i))
```
NIT, like above
```suggestion
      for (i <- 0 <- exceptionLines.length) expect(exceptionLines(i) == expectedLines(i))
```

alexarchambault reviewed (approved):
Looks good to me!

(I only added two NIT comments. Take them into account if you're ok with them, but it's fine merging as is, else…)
